src/reconcile/rec_visibility.c

```diff
@@ -869,26 +978,80 @@ __rec_fill_tw_from_upd_select(
      * that the value is visible to any timestamp/transaction id ahead of it.
      */
     if (upd->type == WT_UPDATE_TOMBSTONE) {
-        WT_TIME_WINDOW_SET_STOP(select_tw, upd, upd->prepare_state);
+        WT_TIME_WINDOW_SET_STOP(select_tw, upd, write_prepare);
         tombstone = upd_select->tombstone = upd;
 
-        /* Find the update this tombstone applies to. */
-        if (!__wt_txn_upd_visible_all(session, upd)) {
-            while (upd->next != NULL && upd->next->txnid == WT_TXN_ABORTED)
-                upd = upd->next;
+        /*
+         * Find the update this tombstone applies to.
+         *
+         * We need to find the full update if the prepared tombstone is rolled back. We resolve
+         * prepare updates recursively from the oldest to the newest. If we write a prepared
+         * tombstone, we may see a prepared update that is rolled back from the same transaction.
+         * Ensure we also write it to the disk in this case.
+         */
```

Comment:
We can add a local variable:

check_prepare = write_prepare;

Then use this variable to skip some check. Then we can turn this to:

```
                if (!write_prepare)
                    continue;

                if (!F_ISSET(S2C(session), WT_CONN_PRESERVE_PREPARED))
                    continue;
```
to

```
          if (!check_prepare)
                  continue;
```